File: sdks/sandbox/kotlin/sandbox/src/main/kotlin/com/alibaba/opensandbox/sandbox/infrastructure/factory/AdapterFactory.kt
/*
 * Copyright 2025 Alibaba Group Holding Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.opensandbox.sandbox.infrastructure.factory

import com.alibaba.opensandbox.sandbox.HttpClientProvider
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxEndpoint
import com.alibaba.opensandbox.sandbox.domain.services.Commands
import com.alibaba.opensandbox.sandbox.domain.services.CredentialVault
import com.alibaba.opensandbox.sandbox.domain.services.Diagnostics
import com.alibaba.opensandbox.sandbox.domain.services.Egress
import com.alibaba.opensandbox.sandbox.domain.services.Filesystem
import com.alibaba.opensandbox.sandbox.domain.services.Health
import com.alibaba.opensandbox.sandbox.domain.services.IsolationService
import com.alibaba.opensandbox.sandbox.domain.services.Metrics
import com.alibaba.opensandbox.sandbox.domain.services.Sandboxes
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.service.CommandsAdapter
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.service.DiagnosticsAdapter
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.service.EgressAdapter
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.service.FilesystemAdapter
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.service.HealthAdapter
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.service.IsolatedSessionsAdapter
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.service.MetricsAdapter
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.service.NetworkPolicyAdapter
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.service.SandboxesAdapter

/**
 * Factory responsible for creating adapter instances.
 *
 * This factory encapsulates the instantiation logic of specific adapters,
 * decoupling the Sandbox domain object from infrastructure implementation details.
 */
internal class AdapterFactory(
    private val httpClientProvider: HttpClientProvider,
) {
    data class EgressStack(
        val egress: Egress,
        val credentialVault: CredentialVault,
    )

    fun createSandboxes(): Sandboxes {
        return SandboxesAdapter(httpClientProvider)
    }

    fun createDiagnostics(): Diagnostics {
        return DiagnosticsAdapter(httpClientProvider)
    }

    fun createFilesystem(endpoint: SandboxEndpoint): Filesystem {
        return FilesystemAdapter(httpClientProvider, endpoint)
    }

    fun createCommands(endpoint: SandboxEndpoint): Commands {
        return CommandsAdapter(httpClientProvider, endpoint)
    }

    fun createEgressStack(endpoint: SandboxEndpoint): EgressStack {
        val adapter = EgressAdapter(httpClientProvider, endpoint)
        return EgressStack(
            egress = adapter,
            credentialVault = adapter,
        )
    }

    /**
     * Builds an egress stack backed by the lifecycle control plane for
     * template-backed sandboxes, which have no sandbox-side egress sidecar.
     */
    fun createNetworkPolicyStack(sandboxId: String): EgressStack {
        val adapter = NetworkPolicyAdapter(httpClientProvider, sandboxId)
        return EgressStack(
            egress = adapter,
            credentialVault = adapter,
        )
    }

    fun createMetrics(endpoint: SandboxEndpoint): Metrics {
        return MetricsAdapter(httpClientProvider, endpoint)
    }

    fun createHealth(endpoint: SandboxEndpoint): Health {
        return HealthAdapter(httpClientProvider, endpoint)
    }

    fun createIsolatedSessions(endpoint: SandboxEndpoint): IsolationService {
        return IsolatedSessionsAdapter(httpClientProvider, endpoint)
    }
}

File: sdks/sandbox/javascript/src/factory/adapterFactory.ts
// Copyright 2026 Alibaba Group Holding Ltd.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type { ConnectionConfig } from "../config/connection.js";
import type { SandboxId } from "../models/sandboxes.js";
import type { SandboxFiles } from "../services/filesystem.js";
import type { CredentialVault, Egress } from "../services/egress.js";
import type { ExecdCommands } from "../services/execdCommands.js";
import type { ExecdHealth } from "../services/execdHealth.js";
import type { ExecdMetrics } from "../services/execdMetrics.js";
import type { IsolationService } from "../services/isolatedSessions.js";
import type { Sandboxes } from "../services/sandboxes.js";

export interface CreateLifecycleStackOptions {
  connectionConfig: ConnectionConfig;
  lifecycleBaseUrl: string;
}

export interface LifecycleStack {
  sandboxes: Sandboxes;
}

export interface CreateExecdStackOptions {
  connectionConfig: ConnectionConfig;
  execdBaseUrl: string;
  endpointHeaders?: Record<string, string>;
}

export interface ExecdStack {
  commands: ExecdCommands;
  files: SandboxFiles;
  health: ExecdHealth;
  metrics: ExecdMetrics;
  isolation?: IsolationService;
}

export interface CreateEgressStackOptions {
  connectionConfig: ConnectionConfig;
  egressBaseUrl: string;
  endpointHeaders?: Record<string, string>;
}

export interface EgressStack {
  egress: Egress;
  credentialVault?: CredentialVault;
}

export interface CreateNetworkPolicyStackOptions {
  connectionConfig: ConnectionConfig;
  lifecycleBaseUrl: string;
  sandboxId: SandboxId;
}

/**
 * Factory abstraction to keep `Sandbox` and `SandboxManager` decoupled from concrete adapter implementations.
 *
 * This is primarily useful for advanced integrations (custom transports, dependency injection, testing).
 */
export interface AdapterFactory {
  createLifecycleStack(opts: CreateLifecycleStackOptions): LifecycleStack;
  createExecdStack(opts: CreateExecdStackOptions): ExecdStack;
  createEgressStack(opts: CreateEgressStackOptions): EgressStack;
  /**
   * Create an egress stack that routes policy operations through the lifecycle
   * control plane (`/sandboxes/{sandboxId}/networkpolicy`) instead of the
   * sandbox-side egress sidecar.
   *
   * Used for template-backed sandboxes, which have no sandbox-side sidecar.
   * Optional: custom adapter factories may omit it, in which case template
   * backed sandboxes fail fast when egress policy operations are attempted.
   */
  createNetworkPolicyStack?(opts: CreateNetworkPolicyStackOptions): EgressStack;
}
